Feature/follow

vectorch9 opened the pull request:
- [x]  팔로우 구현
- [x]  팔로우 목록
- [x]  팔로워 목록
- [x] 시큐리티 설정 수정

Resolves: #33

vectorch9 committed:
Feat: 팔로우를 위한 시큐리티 설정 추가

시큐리티 설정 추가
querydsl을 사용하기 위한 JpaQueryFactory Bean주입

vectorch9 committed:
Feat: 팔로우 구현

팔로우,언팔로우 구현
팔로윙목록, 팔로워목록 조회 구현
관련 예외, ResultCode 추가

vectorch9 committed:
Fix: Todo 주석 삭제

vectorch9 committed:
Fix: FollowRepositorySupport 삭제

사용하지 않는 파일 삭제

vectorch9 commented on src/main/java/cloneproject/Instagram/service/FollowService.java:
```diff
@@ -0,0 +1,119 @@
+        return result;
+    }
+
+    /**
+     * member_id(pk)만을 담은 팔로잉 목록
+     * 메인화면(포스트목록) 구현 때 사용
+     * @return List<Long>: 로그인한 사용자가 팔로우중인 멤버들의 id 목록
+     */
+    @Transactional(readOnly = true)
+    public List<Long> getOnlyFollowingsMemberId(){
+        String memberId = SecurityContextHolder.getContext().getAuthentication().getName();
+        List<Follow> follows = followRepository.findAllByMemberId(Long.valueOf(memberId));
+        List<Long> result = follows.stream()
+                                        .map(follow->follow.getFollowMember().getId())
+                                        .toList();
+        return result;
+
+    }
+
```
주석 달아놓은대로 포스트목록조회 구현하실 때  필요하실 것 같아서 미리 만들어놨습니다!

vectorch9 commented on src/main/java/cloneproject/Instagram/service/FollowService.java:
```diff
@@ -0,0 +1,119 @@
+
+    }
+
+    /**
+     * 프로필, 게시물에서 팔로우 여부를 판단하기 위한 메서드
+     * @param memberId 멤버ID (현재 로그인 중인 사용자의 PK)
+     * @param followMemberUsername 대상의 username
+     * @return boolean: 팔로우 여부
+     */
+    @Transactional(readOnly = true)
+    public boolean isFollowing(String followMemberUsername){
+        String memberId = SecurityContextHolder.getContext().getAuthentication().getName();
+        Member followMember = memberRepository.findByUsername(followMemberUsername)
+                                                .orElseThrow(MemberDoesNotExistException::new);
+        return followRepository.existsByMemberIdAndFollowMemberId(Long.valueOf(memberId), followMember.getId());
+    }
+
```
마찬가지로 팔로우를 안한 상태라면 팔로우 버튼을 표시해주어야 하는 부분이 몇군데있어서 만들어놨습니다

seonpilKim commented on src/main/java/cloneproject/Instagram/service/FollowService.java:
```diff
@@ -0,0 +1,119 @@
+                                                .map(this::convertMemberToUsernameWithImages)
+                                                .toList();
+        return result;
+    }
```
응답 데이터 리스트에 각각 `팔로우 여부` + `회원 닉네임(name)` + `24시간 내 스토리 업로드 여부`도 추가로 필요할 것 같아요!

seonpilKim commented on src/main/java/cloneproject/Instagram/controller/FollowController.java:
```diff
@@ -0,0 +1,57 @@
+
+@RestController
+@RequiredArgsConstructor
+public class FollowController {
+    
+    private final FollowService followService;
+
+    @PostMapping("/{followMemberUsername}/follow")
+    public ResponseEntity<ResultResponse> follow(@PathVariable("followMemberUsername") String followMemberUsername){
+        boolean success = followService.follow(followMemberUsername);
+        
+        ResultResponse result = ResultResponse.of(ResultCode.FOLLOW_SUCCESS, success);
+        return new ResponseEntity<>(result, HttpStatus.valueOf(result.getStatus()));
+    }
+
+    @PostMapping("/{followMemberUsername}/unfollow")
+    public ResponseEntity<ResultResponse> unfollow(@PathVariable("followMemberUsername") String followMemberUsername){
+        boolean success = followService.unfollow(followMemberUsername);
+        
+        ResultResponse result = ResultResponse.of(ResultCode.UNFOLLOW_SUCCESS, success);
+        return new ResponseEntity<>(result, HttpStatus.valueOf(result.getStatus()));
+    }
+
+    @GetMapping("/{memberUsername}/following")
+    public ResponseEntity<ResultResponse> getFollowings(@PathVariable("memberUsername") String memberUsername){
+        List<UsernameWithImage> followings = followService.getFollowings(memberUsername);
+        
+        ResultResponse result = ResultResponse.of(ResultCode.GET_FOLLOWINGS_SUCCESS, followings);
+        return new ResponseEntity<>(result, HttpStatus.valueOf(result.getStatus()));
+    }
+
+    @GetMapping("/{memberUsername}/followers")
+    public ResponseEntity<ResultResponse> getFollowers(@PathVariable("memberUsername") String memberUsername){
+        List<UsernameWithImage> followings = followService.getFollowers(memberUsername);
+        
+        ResultResponse result = ResultResponse.of(ResultCode.GET_FOLLOWERS_SUCCESS, followings);
+        return new ResponseEntity<>(result, HttpStatus.valueOf(result.getStatus()));
+    }
+
+}
```
메소드 인자마다 `@Validated` `@NotBlank(message = "...")` 추가하면 좋을 것 같아요~

seonpilKim reviewed (commented):
고생하셨습니다. 
리뷰에 언급한 부분들은 시험 끝나시고 천천히 추가 해주시면 될 것 같아요~

vectorch9 committed:
Fix: AccessToken의 유효시간 10분으로 조정

토큰만료 테스트를 위해 유효시간을 1분으로 해놨었는데
다시 10분으로 변경

vectorch9 committed:
Fix: Follow Response, Controller 수정

Security 설정: following,followers 목록 조회도 ROLE_USER로 변경
Follow Controller의 parameter에 validate 추가
Follow 응답에 팔로우여부, 이름 추가

vectorch9 committed:
Merge branch 'Feature/Follow' of github.com:Instagram-Clone-Coding/Spring_instagram-clone into Feature/Follow

vectorch9 committed:
Feat: 자기자신을 팔로우하는지 검증하는 로직 추가

자기자신을 팔로우/언팔로우할 수 없도록 검증하는 로직추가
ErrorCode 중복되는 부분 수정
